render credentials sometimes wipes ca.pem on v0.9.9-rc.2

Steps:
* `init` a new cluster render
* Copy `ca.pem` and `ca-key.pem` into the credentials folder for this new cluster
* Run `render credentials`, all creds generated as expected
* Run `render credentials` again, for example after changing some LB config
* `ca.pem` is now empty

Ah damn, this explains it. I wanted to update the config but couldn't figure out why all the commands are failing, I assumed it's something with the config. 

Is there a way to recover access to the cluster?

```
# kube-aws render stack

Generating credentials...
-> Parsing existing TLS CA
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xad9f64]

goroutine 1 [running]:
github.com/kubernetes-incubator/kube-aws/tlsutil.DecodeCertificatePEM(0xc4201c2c00, 0x0, 0x200, 0x0, 0x200, 0x0)
	/go/src/github.com/kubernetes-incubator/kube-aws/tlsutil/pem.go:32 +0x44
github.com/kubernetes-incubator/kube-aws/core/controlplane/config.(*Cluster).NewAssetsOnDisk(0xc42007c380, 0xe69bf8, 0xb, 0x0, 0xe77588, 0x18, 0xe72716, 0x14, 0xc42007c380, 0x0, ...)
	/go/src/github.com/kubernetes-incubator/kube-aws/core/controlplane/config/encrypted_assets.go:168 +0xb58
github.com/kubernetes-incubator/kube-aws/core/root/render.credentialsRendererImpl.RenderCredentials(0xc42007c380, 0x0, 0xe77588, 0x18, 0xe72716, 0x14, 0xc68280, 0x0)
	/go/src/github.com/kubernetes-incubator/kube-aws/core/root/render/credentials.go:30 +0x9f
github.com/kubernetes-incubator/kube-aws/cmd.runCmdRenderCredentials(0x154ce40, 0x1570bf8, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/kubernetes-incubator/kube-aws/cmd/render.go:98 +0x106
github.com/kubernetes-incubator/kube-aws/vendor/github.com/spf13/cobra.(*Command).execute(0x154ce40, 0x1570bf8, 0x0, 0x0, 0x154ce40, 0x1570bf8)
	/go/src/github.com/kubernetes-incubator/kube-aws/vendor/github.com/spf13/cobra/command.go:647 +0x3f8
github.com/kubernetes-incubator/kube-aws/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x154d4a0, 0xc420183f50, 0x0, 0x0)
	/go/src/github.com/kubernetes-incubator/kube-aws/vendor/github.com/spf13/cobra/command.go:726 +0x339
github.com/kubernetes-incubator/kube-aws/vendor/github.com/spf13/cobra.(*Command).Execute(0x154d4a0, 0xc4200001a0, 0xc4200001a0)
	/go/src/github.com/kubernetes-incubator/kube-aws/vendor/github.com/spf13/cobra/command.go:685 +0x2b
main.main()
	/go/src/github.com/kubernetes-incubator/kube-aws/main.go:10 +0x2d

# kube-aws validate --s3-uri s3://XXXXX                                                     

Error: Failed to initialize cluster driver: failed to render controller cloud config: cloud-config validation errors:
line 595: error: content cannot be decoded as "gzip+base64"
line 595: error: content cannot be decoded as "gzip+base64"
line 595: error: content cannot be decoded as "gzip+base64"
line 595: warning: incorrect type for "content" (want string)
line 595: warning: incorrect type for "content" (want string)
line 595: warning: incorrect type for "content" (want string)
```

--- a/core/controlplane/config/credential.go
+++ b/core/controlplane/config/credential.go
@@ -80,7 +80,10 @@ func EncryptedCredentialCacheFromRawCredential(raw *RawCredentialOnDisk, bytesEn
 }
 
 func RawCredentialFileFromPath(filePath string, defaultValue *string) (*RawCredentialOnDisk, error) {
-	if _, err := os.Stat(filePath); os.IsNotExist(err) && defaultValue != nil {
+	if _, err := os.Stat(filePath); os.IsNotExist(err) {
+		if defaultValue == nil {
+			return nil, fmt.Errorf("%s must exist. Please confirm that you have not deleted the file manually", filePath)
+		}
 		if err := ioutil.WriteFile(filePath, []byte(*defaultValue), 0644); err != nil {
 			return nil, err
 		}
@@ -102,6 +105,9 @@ func (c *RawCredentialOnDisk) Fingerprint() string {
 }
 
 func (c *RawCredentialOnDisk) Persist() error {
+	if len(c.content) == 0 {
+		return fmt.Errorf("%s is going to be empty. Maybe a bug", c.filePath)
+	}
 	if err := ioutil.WriteFile(c.filePath, c.content, 0600); err != nil {
 		return err
 	}

--- a/core/controlplane/config/encrypted_assets.go
+++ b/core/controlplane/config/encrypted_assets.go
@@ -440,39 +440,46 @@ func ReadOrEncryptAssets(dirname string, manageCertificates bool, caKeyRequiredO
 }
 
 func (r *RawAssetsOnMemory) WriteToDir(dirname string, includeCAKey bool) error {
+	workerCAKeyDefaultSymlinkTo := ""
+	if includeCAKey {
+		workerCAKeyDefaultSymlinkTo = "ca-key.pem"
+	}
+
 	assets := []struct {
-		name      string
-		data      []byte
-		overwrite bool
+		name             string
+		data             []byte
+		overwrite        bool
+		ifEmptySymlinkTo string
 	}{
-		{"ca.pem", r.CACert, true},
-		{"worker-ca.pem", r.WorkerCACert, true},
-		{"worker-ca-key.pem", r.WorkerCAKey, true},
-		{"apiserver.pem", r.APIServerCert, true},
-		{"apiserver-key.pem", r.APIServerKey, true},
-		{"worker.pem", r.WorkerCert, true},
-		{"worker-key.pem", r.WorkerKey, true},
-		{"admin.pem", r.AdminCert, true},
-		{"admin-key.pem", r.AdminKey, true},
-		{"etcd.pem", r.EtcdCert, true},
-		{"etcd-key.pem", r.EtcdKey, true},
-		{"etcd-client.pem", r.EtcdClientCert, true},
-		{"etcd-client-key.pem", r.EtcdClientKey, true},
-		{"etcd-trusted-ca.pem", r.EtcdTrustedCA, true},
-		{"kubelet-tls-bootstrap-token", r.TLSBootstrapToken, true},
+		{"ca.pem", r.CACert, true, ""},
+		{"worker-ca.pem", r.WorkerCACert, true, "ca.pem"},
+		{"worker-ca-key.pem", r.WorkerCAKey, true, workerCAKeyDefaultSymlinkTo},
+		{"apiserver.pem", r.APIServerCert, true, ""},
+		{"apiserver-key.pem", r.APIServerKey, true, ""},
+		{"worker.pem", r.WorkerCert, true, ""},
+		{"worker-key.pem", r.WorkerKey, true, ""},
+		{"admin.pem", r.AdminCert, true, ""},
+		{"admin-key.pem", r.AdminKey, true, ""},
+		{"etcd.pem", r.EtcdCert, true, ""},
+		{"etcd-key.pem", r.EtcdKey, true, ""},
+		{"etcd-client.pem", r.EtcdClientCert, true, ""},
+		{"etcd-client-key.pem", r.EtcdClientKey, true, ""},
+		{"etcd-trusted-ca.pem", r.EtcdTrustedCA, true, "ca.pem"},
+		{"kubelet-tls-bootstrap-token", r.TLSBootstrapToken, true, ""},
 
 		// Content entirely provided by user, so do not overwrite it if
 		// the file already exists
-		{"tokens.csv", r.AuthTokens, false},
+		{"tokens.csv", r.AuthTokens, false, ""},
 	}
 
 	if includeCAKey {
 		// This is required to be linked from worker-ca-key.pem
 		assets = append(assets, struct {
-			name      string
-			data      []byte
-			overwrite bool
-		}{"ca-key.pem", r.CAKey, true})
+			name             string
+			data             []byte
+			overwrite        bool
+			ifEmptySymlinkTo string
+		}{"ca-key.pem", r.CAKey, true, ""})
 	}
 
 	for _, asset := range assets {
@@ -489,55 +496,62 @@ func (r *RawAssetsOnMemory) WriteToDir(dirname string, includeCAKey bool) error 
 				return err
 			}
 		}
+		if len(asset.data) == 0 {
+			if asset.ifEmptySymlinkTo != "" {
+				// etcd trusted ca and worker-ca are separate files, but pointing to ca.pem by default.
+				// In advanced configurations, when certs are managed outside of kube-aws,
+				// these can be separate CAs to ensure that worker nodes have no certs which would let them
+				// access etcd directly. If worker-ca.pem != ca.pem, then ca.pem should include worker-ca.pem
+				// to let TLS bootstrapped workers acces APIServer.
+				wd, err := os.Getwd()
+				if err != nil {
+					return err
+				}
+
+				if err := os.Chdir(dirname); err != nil {
+					return err
+				}
+
+				// The path of the symlink
+				from := asset.name
+				// The path to the actual file
+				to := asset.ifEmptySymlinkTo
+
+				lstatFileInfo, lstatErr := os.Lstat(from)
+				symlinkExists := lstatErr == nil && (lstatFileInfo.Mode()&os.ModeSymlink == os.ModeSymlink)
+				fileExists := lstatErr == nil && !symlinkExists
+
+				if fileExists {
+					fmt.Printf("INFO: Removing a file at %s\n", from)
+					if err := os.Remove(from); err != nil {
+						return err
+					}
+				}
+
+				if symlinkExists {
+					fmt.Printf("INFO: Removing a symlink at %s\n", from)
+					if err := os.Remove(from); err != nil {
+						return err
+					}
+				}
+
+				fmt.Printf("INFO: Creating a symlink from %s to %s\n", from, to)
+				if err := os.Symlink(to, from); err != nil {
+					return err
+				}
+
+				if err := os.Chdir(wd); err != nil {
+					return err
+				}
+				continue
+			} else if asset.name != "tokens.csv" {
+				return fmt.Errorf("Not sure what to do for %s", path)
+			}
+		}
+		fmt.Printf("INFO: Writing %d bytes to %s\n", len(asset.data), path)
 		if err := ioutil.WriteFile(path, asset.data, 0600); err != nil {
 			return err
 		}
-	}
-
-	// etcd trusted ca and worker-ca are separate files, but pointing to ca.pem by default.
-	// In advanced configurations, when certs are managed outside of kube-aws,
-	// these can be separate CAs to ensure that worker nodes have no certs which would let them
-	// access etcd directly. If worker-ca.pem != ca.pem, then ca.pem should include worker-ca.pem
-	// to let TLS bootstrapped workers acces APIServer.
-	type symlink struct {
-		from string
-		to   string
-	}
-	symlinks := []symlink{
-		{"ca.pem", "worker-ca.pem"},
-		{"ca.pem", "etcd-trusted-ca.pem"},
-	}
-
-	if includeCAKey {
-		symlinks = append(symlinks, symlink{"ca-key.pem", "worker-ca-key.pem"})
-	}
-
-	wd, err := os.Getwd()
-	if err != nil {
-		return err
-	}
-
-	if err := os.Chdir(dirname); err != nil {
-		return err
-	}
-
-	for _, sl := range symlinks {
-		from := sl.from
-		to := sl.to
-
-		if _, err := os.Lstat(to); err == nil {
-			if err := os.Remove(to); err != nil {
-				return err
-			}
-		}
-
-		if err := os.Symlink(from, to); err != nil {
-			return err
-		}
-	}
-
-	if err := os.Chdir(wd); err != nil {
-		return err
 	}
 
 	return nil

